socket_wrapper.cpp

```diff
@@ -30,43 +30,60 @@ void SocketWrapper::InitSocketAddress(int family, int address, int port)
     m_socket_address.sin_port = port;
 }
 
-void SocketWrapper::SetInAddressWithStr(const char* ip_address_str)
+void Socket::SetInAddressWithStr(const char* ip_address_str)
 {
     if((inet_pton(AF_INET, ip_address_str, &m_socket_address.sin_addr)) <= 0)
-        throw socket_exception("inet_pton returned error");
+        throw SocketException("inet_pton returned error");
 }
 
-void SocketWrapper::Bind()
+void Socket::Bind(const unsigned address, const unsigned port)
 {
+    m_socket_address.sin_addr.s_addr = address;
+    m_socket_address.sin_port = port;
+
     if(bind(m_socket_handle, (const sockaddr*)&m_socket_address, sizeof(m_socket_address)) < 0)
-        throw socket_exception("bind returned error");
+        throw SocketException("bind returned error");
 }
 
-void SocketWrapper::Listen(int flag)
+void Socket::Listen(int flag)
 {
     if(listen(m_socket_handle, flag) < 0)
-        throw socket_exception("listen returned error");
+        throw SocketException("listen returned error");
 }
 
-std::unique_ptr<SocketWrapper> SocketWrapper::Accept()
+Socket_sptr Socket::Accept()
 {
     struct sockaddr_in client_address;
 	socklen_t client_address_len = 0;
     int client_socket_handle = accept(m_socket_handle, (sockaddr*)&client_address, &client_address_len);
-    if(client_socket_handle == -1) throw socket_exception("accept returned error");
-    return std::make_unique<SocketWrapper>(client_socket_handle, (sockaddr*)&client_address);
+    if(client_socket_handle == -1) throw SocketException("accept returned error");
+    return std::make_shared<Socket>(client_socket_handle, (sockaddr*)&client_address);
 }
 
-void SocketWrapper::Close()
+void Socket::Close()
 {
     if(close(m_socket_handle) < 0)
-        throw socket_exception("close returned error");
+        throw SocketException("close returned error");
 }
 
-void SocketWrapper::Send(const void* buffer, size_t buffer_length, int flags)
+void Socket::Send(const void* buffer, size_t buffer_length, int flags)
 {
     if(send(m_socket_handle, buffer, buffer_length, flags) < 0)
-       throw socket_exception("send returned error");
+       throw SocketException("send returned error");
+}
+
+int Socket::Recv(void* buffer, size_t buffer_length, int flags)
```

Comment:
Both of `recv` and `read` can return `-1` in case of error. I'm sure we want to throw an exception in both cases.
We'll need to store return value to temporary variable, so lets make assignment and check as separate code lines, 'cause it is more clean way.